components/scheduler2/generator/executors.go

```diff
@@ -0,0 +1,289 @@
+		// conflict here (after transitioning Executor), that's fine; buffered actions
+		// are idempotent through their generated request IDs
+		return hsm.MachineTransition(generatorNode, func(g Generator) (hsm.TransitionOutput, error) {
+			if g.ConflictToken != generator.ConflictToken {
+				return hsm.TransitionOutput{}, fmt.Errorf(
+					"%w: conflicting Generator state while buffering actions",
+					consts.ErrStaleState,
+				)
+			}
+
+			g.GeneratorInternalState = generator.GeneratorInternalState
+			g.ConflictToken++
+			return TransitionSleep.Apply(g, EventSleep{
+				Node:     generatorNode,
+				Deadline: generator.NextInvocationTime.AsTime(),
+			})
+		})
+	})
+}
+
+func (e taskExecutor) logSchedule(msg string, scheduler core.Scheduler) {
+	// log spec as json since it's more readable than the Go representation
+	specJson, _ := protojson.Marshal(scheduler.Schedule.Spec)
+	policiesJson, _ := protojson.Marshal(scheduler.Schedule.Policies)
+	e.Logger.Info(msg,
+		tag.NewStringTag("spec", string(specJson)),
+		tag.NewStringTag("policies", string(policiesJson)))
+}
+
+type processedTimeRangeResult struct {
+	NextWakeupTime time.Time
+	LastActionTime time.Time
+	BufferedStarts []*schedpb.BufferedStart
+}
+
+// Processes the given time range, generating buffered actions according to the
+// schedule spec.
+func (e taskExecutor) processTimeRange(
+	scheduler core.Scheduler,
+	tweakables common.Tweakables,
+	start, end time.Time,
+	limit *int,
+) (*processedTimeRangeResult, error) {
+	overlapPolicy := scheduler.OverlapPolicy()
+
+	e.Logger.Debug("processTimeRange",
+		tag.NewTimeTag("start", start),
+		tag.NewTimeTag("end", end),
+		tag.NewAnyTag("overlap-policy", overlapPolicy),
+		tag.NewBoolTag("manual", false))
+
+	catchupWindow := e.catchupWindow(scheduler, tweakables)
+
+	// Peek at paused/remaining actions state and don't bother if we're not going to
+	// take an action now. (Don't count as missed catchup window either.)
+	// Skip over entire time range if paused or no actions can be taken.
+	if !scheduler.CanTakeScheduledAction(false) {
+		// use end as last action time so that we don't reprocess time spent paused
+		next, err := e.getNextTime(scheduler, end)
+		if err != nil {
+			return nil, err
+		}
+
+		return &processedTimeRangeResult{
+			NextWakeupTime: next.Next,
+			LastActionTime: end,
+			BufferedStarts: nil,
+		}, nil
+	}
+
+	lastAction := start
+	var next scheduler1.GetNextTimeResult
+	var bufferedStarts []*schedpb.BufferedStart
+	for next, err := e.getNextTime(scheduler, start); err == nil && !(next.Next.IsZero() || next.Next.After(end)); next, err = e.getNextTime(scheduler, next.Next) {
+		if scheduler.Info.UpdateTime.AsTime().After(next.Next) {
+			// We're reprocessing since the most recent event after an update. Discard actions before
+			// the update time (which was just set to "now"). This doesn't have to be guarded with
+			// hasMinVersion because this condition couldn't happen in previous versions.
```

Comment:
I imagine this comment should be rewritten somewhat. We don't have versions anymore, at least. Also I'm not sure how updates fit into this execution model, it may not be true that "we just set update time to 'now'"

Reply:
Oops, yes, will rewrite this.